### ISSUE TITLE
[SUGGESTION] I think I've gone mad, but...

### ISSUE THREAD

**Issue by TheK0tYaRa** [cheat-engine/cheat-engine#1111]
...I'm gonna try to make CE work on Ubuntu LTS 18.04.3
If I will fail at any moment I won't go asking for help, but if it will work successfully - it will be a nice experience and will be put to public judgement. Latest FPC for this Ubuntu is 3.0.4. It may be a point of failure, but a problem is always a reason to find or make a solution! No promises though.

**Comment by cheat-engine** [comment]
Good luck.  Which widgetset will you use GTK or QT ?

Also, posrting isn't that difficult, just replace the windows unit with a selfwritten linux one that implements the missing functions (like OpenProcess, Toolhelp32snapshot, etc...)   But it's similar to ceserver but then just local (and in pascal, though I guess you could also write it in C and then make CE link to the .o file )

**Comment by Symbai** [comment]
@cheat-engine: What are the linux equivalents for VirtualAllocEx and VirtualQueryEx?

**Comment by cheat-engine** [comment]
for virtualallocex i usually go for suspend, change rip to point to code that allocates memory, run till inf loop and restore register state

for virtualqueryex just textparse /proc/pid/maps

**Comment by Symbai** [comment]
Why not using mmap for allocation? Like in CEServer https://github.com/cheat-engine/cheat-engine/blob/acc6cefd0f18d9dd97040e6cc0a5339e9f056e2e/Cheat%20Engine/ceserver/extension/server.c#L234 or is this only for local proces like VirtualAlloc (no EX) in Windows?

For read/write I found process_vm_readv process_vm_writev to be working extremely well. Sadly i fails to write to a module. I believe its because of the memory page protection similar to Windows.

Do you know the equavilent of VirtualProtectEx?

**Comment by cheat-engine** [comment]
extension/server.c runs inside the target process.  It uses the same method as described above to call dlopen to load the extension.so

VirtualProtectEx would be mprotect?

Again, kernel modules are a lot easier and can bypass some protections like the pkey stuff for mprotect if it's used

**Comment by Symbai** [comment]
That's what I thought too, but I tried mprotect but it always fails with errno code 12 (out of memory?!). I use the same address as for process_vm_readv, the size of the memory region and READ | WRITE | EXEC flag as parameters.

**Comment by cheat-engine** [comment]
page table entry editing then, or get the physical address of the memory and change it there

**Comment by Symbai** [comment]
Yup `process_vm_writev` fails writing to modules. Using ptrace with poke_data works out of the box. It's kinda frustrating that ptrace works but process_vm_writev does not, as ptrace requires the debugger to attach which fails very easily, because only one debugger can attach to the game at the same time and some games attach to itself as kind of anti-cheat.

Then there is `pread64` / `pwrite64` but I don't know how I can share the game memory to get the required file descriptor parameter and I dont know if this requires ptrace to be attached as well. Is that what you meant with page table entry editing? Reading the /proc/pid/mem ?

**Comment by Symbai** [comment]
Okay figured it out. pwrite64 works as well writing to the module, without using ptrace.

**Comment by Symbai** [comment]
When trying to port small portion of CE to Linux I'm running into `Segmentation fault` exception on lines which uses `setlength()`. @cheat-engine: Do you know if `setlength` is compatible on Linux or something different when targeting linux? I couldn't really find anything using Google except of https://forum.lazarus.freepascal.org/index.php?topic=28059.0 which doesn't really provide any solution.

Reproduce:
```
type TMyByteArray=array of byte;

procedure Hello();
var 
    b: TMyByteArray
begin
    Bla(b);
end;

procedure Bla(var bytes: TMyByteArray);
begin
    setlength(bytes,0); // <---- segment fault
end; 
```

Compiling under Ubuntu 64bit with Lararus 2.0.2:
![image](https://user-images.githubusercontent.com/14368203/73600432-f5612380-4547-11ea-8b24-a781440fd7ce.png)

**Comment by PopeRigby** [comment]
@TheK0tYaRa Did you ever make any progress on this?
